ingestor/chalicelib/yankee.py

```diff
@@ -0,0 +1,285 @@
+from typing import Dict, Optional, Tuple
+import requests
+import time
+import pandas as pd
+import json
+import boto3
+from keys import YANKEE_API_KEY
+from botocore.exceptions import ClientError
+from datetime import datetime
+from tempfile import TemporaryDirectory
+import s3
+from ddtrace import tracer
+
+from typing import List
+from sqlalchemy.orm import Session
+from mbta_gtfs_sqlite import MbtaGtfsArchive
+from mbta_gtfs_sqlite.models import (
+    RoutePattern,
+    RoutePatternTypicality,
+    Trip,
+    ShapePoint,
+)
+
+ShapeDict = Dict[str, List[ShapePoint]]
+
+BUCKET = "tm-shuttle-positions"
+KEY = "yankee/last_shuttle_positions.csv"
+BOSTON_COORDS = (-71.057083, 42.361145)
+OSRM_DISTANCE_API = "http://router.project-osrm.org/route/v1/driving/"
+METERS_PER_MILE = 0.000621371
+SHUTTLE_PREFIX = "Shuttle"
+
+def load_bus_positions():
+    try:
+        data = s3.download(BUCKET, KEY, compressed=False)
+        return data
+    except ClientError as ex:
+        if ex.response["Error"]["Code"] != "NoSuchKey":
+            raise
+
+def get_shuttle_shapes(
+    session: Session,
+) -> ShapeDict:
+    route_patterns = (
+        session.query(RoutePattern)
+        .filter(
+            RoutePattern.route_pattern_typicality == RoutePatternTypicality.DIVERSION,
+            RoutePattern.route_pattern_id.startswith(SHUTTLE_PREFIX)
+        ).all()
+    )
+
+    print(f"Found {len(route_patterns)} active shuttle route patterns")
+
+    shuttle_shapes: ShapeDict = {}
+    for route_pattern in route_patterns:
+        if route_pattern == None:
+            print(f"Unable to fetch route patttern for route id {route_pattern.route_id}")
+            continue
+
+        representative_trip = (
+            session.query(Trip)
+            .filter(Trip.trip_id == route_pattern.representative_trip_id)
+            .first()
+        )
+        #
+        if representative_trip == None:
+            print(f"Unable to fetch route patttern for route id {route_pattern.route_id}")
+            continue
+
+        print(f"Getting shapes for {route_pattern.route_id}")
+
+        shape_points = (
+            session.query(ShapePoint)
+            .filter(ShapePoint.shape_id == representative_trip.shape_id)
+            .order_by(ShapePoint.shape_pt_sequence)
+        ).all()
+
+        shuttle_shapes[route_pattern.route_id] = shape_points
+
+
+    return shuttle_shapes
+
+
+def get_session_for_latest_feed() -> Session:
+
+    s3 = boto3.resource("s3")
+    archive = MbtaGtfsArchive(
+            local_archive_path=TemporaryDirectory().name)
+            #s3_bucket=s3.Bucket("tm-gtfs"))
```

Comment:
this will be uncommented later. Question: could we have a separate lambda for updating the GTFS feed less frequently? Maybe on a daily basis?